File: src/algorithms/python/Insert_Interval.py
# -*- coding: utf-8 -*-


# Definition for an interval.
class Interval:
    def __init__(self, s=0, e=0):
        self.start = s
        self.end = e


class Solution:
    # @param {Interval[]} intervals
    # @param {Interval} newInterval
    # @return {Interval[]}
    def insert(self, intervals, newInterval):
        rec, ret = [], []
        intervals.append(newInterval)
        for interval in intervals:
            rec.append((interval.start, interval.end))
        rec.sort()
        s, e = rec[0][0], rec[0][1]
        for t in rec[1:]:
            if t[0] <= e:
                e = max(t[1], e)
            else:
                ret.append(Interval(s, e))
                s, e = t[0], t[1]
        ret.append(Interval(s, e))
        return ret


if __name__ == '__main__':
    solution = Solution()
    a, b = Interval(1, 2), Interval(3, 5)
    c, d = Interval(6, 7), Interval(8, 10)
    e, f = Interval(12, 16), Interval(4, 9)
    intervals = solution.insert([a, b, c, d, e], f)
    for interval in intervals:
        print interval.start, interval.end

File: src/algorithms/cpp/Insert_Interval.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

struct Interval {
    int start;
    int end;
    Interval() : start(0), end(0) {}
    Interval(int s, int e) : start(s), end(e) {}
};

bool cmp(Interval a, Interval b) {
    return a.start < b.start;
}


class Solution {
public:
    vector <Interval> insert(vector <Interval>& intervals, Interval newInterval) {
        intervals.push_back(newInterval);
        vector <Interval> result;
        sort(intervals.begin(), intervals.end(), cmp);
        int start = intervals[0].start, end = intervals[0].end;
        for (int i = 1; i < SZ(intervals); ++ i) {
            if (start <= intervals[i].start && intervals[i].start <= end) {
                get_max(end, intervals[i].end);
            }
            else {
                result.push_back(Interval(start, end));
                start = intervals[i].start, end = intervals[i].end;
            }
        }
        result.push_back(Interval(start, end));
        return result;
    }
};

int main() {
    return 0;
}
